concurrency/MemberReceiver.java

```diff
@@ -0,0 +1,2351 @@
+
+    if (timeCurrentPacketReceived == 0) {
+        rtpElapsed = 0;
+    } else {
+        rtpElapsed = (System.currentTimeMillis() - timeCurrentPacketReceived) / 1000;
+    }
+
+    long rtcpElapsed = rtcpReceiver.secondsSinceLastReport(member.getRtcpAddress());
+
+    if (rtcpElapsed < RtpSocket.getRtpTimeout() || rtpElapsed < RtpSocket.getRtpTimeout()) {
+        return false;
+    }
+
+    //if (Logger.logLevel >= Logger.LOG_INFO) {
+            Logger.println("Call " + cp
+                + " time since last RTCP report " + rtcpElapsed
+        + " time since last RTP packet received " + rtpElapsed);
+        //}
+
+    /*
+         * We have not received an RTP or RTCP packet in quite some
+         * time.  Assume the call is dead.
+     *
+     * XXX There is a gateway (10.6.4.61)
+     * which doesn't send RTCP packets.
+     * For now, we'll only timeout calls with "sip:" in
+     * the phone number.
+         */
+        Logger.println("Call " + cp
+            + ":  Timeout, cancelling the call...");
+        callHandler.cancelRequest("call timeout, no keepalive received");
+    callCancelled = true;
+
+    return true;
+    }
+
+    /**
+     * Handle data which the ConferenceReceiver has sent to us and
+     * add it to the current whisper group mix.
+     *
+     * There is a single ConferenceReceiver to which all CallParticipants
+     * send data.  The ConferenceReceiver dispatches data to
+     * the appropriate conference member by calling our receive method.
+     *
+     * The job of the conference member is to receive packets from
+     * the call participant and add them to the appropriate whisper group.
+     *
+     * Each MemberReceiver keeps it's own list of data it has contributed
+     * to the mix's linearMixBuffers.
+     *
+     * When a member receives data from the CallParticipant, the member
+     * adds an element of int[] to its own list.  The member then adds
+     * the data to the current whisper group.  The element index
+     * of the member's list is the same index used for the whisper group's
+     * linearMixBuffers.
+     *
+     * Adding and removing list elements is done synchronized on whisperGroup.
+     */
+    public void setDropPackets(int dropPackets) {
+    this.dropPackets = dropPackets;
+    }
+
+    /*
+     * With deferMixing set to true, when a packet arrives it is inserted
+     * in the jitter buffer and mixing is done in saveCurrentContribution().
+     * It may be better to do the mixing when the packet is received so
+     * the work is done by a thread separate from the sender thread.
+     */
+    private static boolean deferMixing = false;
+    private int forcedToDeferMixing;
+
+    public static void deferMixing(boolean deferMixing) {
+    MemberReceiver.deferMixing = deferMixing;
+    }
+
+    public static boolean deferMixing() {
+    return deferMixing;
+    }
+
+    private void forwardData(int[] data) {
+    for (MemberSender memberSender : forwardMemberList) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **953**. At this line the following field is not protected by synchronization: `forwardMemberList`. Your code typically protects this field with synchronization on object: `forwardMemberList`.

**Fix**  
 Consider protecting the above field with synchronization on object: `forwardMemberList`. For example, like at lines: **540**, **312**, **314**, **529**, and **534**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **312**, **314**, **529**, **534**, and **540**.